DOCSP-43072: GridFS (#60)
mongodb/docs-c: source/write/gridfs.txt

   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: binary large object, blob, storage, code example

Overview
--------

In this guide, you can learn how to store and retrieve large files in
MongoDB by using **GridFS**. GridFS is a specification implemented by 
the {+driver-short+} that describes how to split files into chunks when storing them
and reassemble them when retrieving them. The driver's implementation of
GridFS is an abstraction that manages the operations and organization of
the file storage. 

Use GridFS if the size of your files exceeds the BSON document
size limit of 16MB. For more detailed information on whether GridFS is
suitable for your use case, see :manual:`GridFS </core/gridfs>` in the
{+mdb-server+} manual.

How GridFS Works
----------------

GridFS organizes files in a **bucket**, a group of MongoDB collections
that contain the chunks of files and information describing them. The
bucket contains the following collections, named using the convention
defined in the GridFS specification:

- The ``chunks`` collection stores the binary file chunks.
- The ``files`` collection stores the file metadata.

> When you create a new GridFS bucket, the driver creates the preceding 
> collections, prefixed with the default bucket name ``fs``, unless
> you specify a different name. The driver also creates an index on each
> collection to ensure efficient retrieval of the files and related
> metadata. The driver creates the GridFS bucket, if it doesn't exist, only when the first write
> operation is performed. The driver creates indexes only if they don't exist and when the
> bucket is empty. For more information about
> GridFS indexes, see :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
> in the {+mdb-server+} manual.

Reviewer: S: this reads a little more clearly to me

```suggestion
The driver creates the GridFS bucket, if it doesn't exist, when the first write
operation is performed. The bucket contains the preceding 
collections, which it prefixex with the default bucket name ``fs`` unless
you specify a different name. To ensure efficient retrieval of the files and related
metadata, the driver also creates an index on each
collection if they don't exist and when the
bucket is empty. 

For more information about
GridFS indexes, see :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
in the {+mdb-server+} manual.
```

When using GridFS to store files, the driver splits the files into smaller
chunks, each represented by a separate document in the ``chunks`` collection.
It also creates a document in the ``files`` collection that contains
a file ID, file name, and other file metadata. You can upload the file by passing
a stream to the {+driver-short+} to consume or creating a new stream and writing to it
directly.

> View the following diagram to see how GridFS splits the files when uploaded to
> a bucket:

Reviewer: S:

```suggestion
The following diagram shows how GridFS splits files when they are
uploaded to a bucket:
```

.. figure:: /includes/figures/GridFS-upload.png
   :alt: A diagram that shows how GridFS uploads a file to a bucket

> When retrieving files, GridFS fetches the metadata from the ``files``

Reviewer: S:

```suggestion
When you retrieve files from GridFS, it fetches the metadata from the ``files``
```

collection in the specified bucket and uses the information to reconstruct
the file from documents in the ``chunks`` collection. You can read the file
by writing its contents to an existing stream or creating a new stream that points
to the file.

.. _gridfs-create-bucket:

Create a GridFS Bucket
----------------------

> To store or retrieve files from GridFS, call the ``mongoc_gridfs_bucket_new()``

Reviewer: Q: The previous section says the driver creates the bucket automatically when you try to write to it, but it seems like this function is creating it...?

Author: I'll clarify that - this function either creates a new bucket or accesses an existing one. If one doesn't exist, then the new bucket isn't actually created until you perform a write operation

> To store or retrieve files from GridFS, call the ``mongoc_gridfs_bucket_new()``

Reviewer: ```suggestion
To use GridFS, first call the ``mongoc_gridfs_bucket_new()``
```

> function. This function creates a new ``mongoc_gridfs_bucket_t`` value and accepts the

Reviewer: ```suggestion
function. This function creates a new ``mongoc_gridfs_bucket_t`` structure and accepts the
```

(i think)

Author: looks like structure was used in other PRs, so I'll update

   :start-after: start-open-upload-stream
   :end-before: end-open-upload-stream

Upload an Existing Stream
~~~~~~~~~~~~~~~~~~~~~~~~~

Use the ``mongoc_gridfs_bucket_upload_from_stream()`` function to upload the contents
of a stream to a new GridFS file. The ``mongoc_gridfs_bucket_upload_from_stream()``
function allows you to specify configuration information in an options document, which
you can pass as a parameter.

This example performs the following actions:

- Calls the ``mongoc_stream_file_new_for_path()`` function to open a file located
  at ``/path/to/input_file`` as a stream in read only (``O_RDONLY``) mode
- Calls the ``mongoc_gridfs_bucket_upload_from_stream()`` function to upload the
  contents of the stream to a GridFS file named ``"new_file"`` 
- Calls the ``mongoc_stream_close()`` and ``mongoc_stream_destroy()`` functions to
  close and destroy the stream

.. literalinclude:: /includes/write/gridfs.c
   :language: c
   :dedent:
   :start-after: start-upload-from-stream
   :end-before: end-upload-from-stream

.. _gridfs-retrieve-file-info:

Retrieve File Information
-------------------------

In this section, you can learn how to retrieve file metadata stored in the

> ``files`` collection of the GridFS bucket. The metadata contains information

Reviewer: ```suggestion
``files`` collection of the GridFS bucket. A file's metadata contains information
```

about the file it refers to, including:

- The ``_id`` of the file
- The name of the file
- The length/size of the file
- The upload date and time
- A ``metadata`` document in which you can store any other information

To retrieve files from a GridFS bucket, call the ``mongoc_gridfs_bucket_find()``
function and pass your bucket as a parameter. The function returns a cursor
from which you can access the results. To learn more about cursors in the

> {+driver-short+}, see the :ref:`<c-cursors>` guide.

Reviewer: S: put link in an admonition, or at least its own paragraph
